### PR TITLE
Add `shell`-configuration to pane layer

### LINKED ISSUE #139: Respawning panes and remain-on-exit
Using YAML, is there a way to respawn a pane and run the commands for that pane again?

I've tried:

```
session_name: respawnable
windows:
- window_name: watchers
  options:
    remain-on-exit: on
  layout: tiled
  panes:
    - shell_command:
        - ls
    - shell_command:
        - ls
```

I was expecting when I sent: `respawn-pane -k` for "`ls`" to reload. This is not the case.

Thanks.

### PR DESCRIPTION
**Pull request by jerri**
This is supposed to fix #139.

I added a shell-configuration to the pane layer. This way you can specify for every pane what shell/application is supposed to be started. I added an example to the documentation to drive the point home why it makes sense.

I wasn't really able to understand how to add the test cases to the test environment. I hope someone can help with this. The example should hopefully already contain most edge cases in any case.

The more complicated part was to make sure that the shell command for the first pane gets respected by the window creation, as the shell for the first pane has to be handed over to the create-window-command.

### CODE REVIEW

**Comment by jerri** [comment]
Is there anything I can do, to further this merge requests? I am struggling with the test environment. Maybe someone can give me a hand or a hint.

**Comment by tony** [comment]
@jerri can you rebase this?  Sorry for the delay!

Have you seen the developer documentation here?   https://tmuxp.git-pull.com/developing.html

> Maybe someone can give me a hand or a hint.

Are you able to run the tests locally?  What part of this are you having issues with? e.g. writing a new test?

File: examples/pane-shell.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+      - shell: /usr/bin/vim -u none
+        shell_command:
+          - iAll panes have the `remain-on-exit` setting on.
+          - When you exit out of the shell or application, the panes will remain.
+          - Use tmux command `:kill-pane` to remove the pane.
+          - Use tmux command `:respawn-pane` to restart the shell in the pane.
+          - Use <Escape> and then `:q!` to get out of this vim window. :-)
+      - shell_command:
+          - print('Hello World 2')
+      - shell: /usr/bin/top
```
Comment (tony) [inline examples/pane-shell.yaml:22]: thanks for providing an example!

**Comment by jerri** [comment]
I didn't have an idea how to add an appropriate automatic check. On top the automatic tests don't worked on my mac. I found reasons why, but not how to fix them. Added the information to another ticket, that also mentioned the problems.
I can do the rebase. Sure.

**Comment by jerri** [comment]
I fixed the conflict. And push the new version.
I also found my comment in https://github.com/tmux-python/tmuxp/issues/620#issuecomment-829565231 about the problem I had with the unit testing. When commenting out the mentioned test, I could complete the whole test procedure. Unfortunately I sort of ran out of time and couldn't continue on understanding how the test environment works and how to add automatic testing for my change.

**Comment by tony** [comment]
@jerri Hi, sorry for the delay!

Can you rebase / solve conflicts?  (Interestingly I'm not seeing the tests show up)

**Comment by jerri** [comment]
There shouldn't be any more conflict, no? At least I don't see any conflicts mentioned in this pull request.

**Comment by tony** [comment]
@jerri Hi there!

If you rebase against master, you may have conflicts to solve. But right now (and back then) it wasn't rebased.

I'm not familiar with how to understand merge commits like https://github.com/tmux-python/tmuxp/pull/672/commits/c43079c88f64393d49a215bc4ba591e061a0f967. As you can see it's a few pages long.

https://github.com/tmux-python/tmuxp/pull/672/commits

Do you know how to rebase against master? Can I provide a tutorial maybe?

In other words, it would be just your changes as commits, I'm not sure what to make of  the commit with "Merge branch 'master' into issue-139".

**Comment by tony** [comment]
I remember now.  When I asked that question earlier, I was of the belief that GitHub actions tests weren't running because of it.

I suppose GitHub's interface solves a lot of it (I only see the diff) - I'd still prefer having a clean PR. Is that possible? (Not via merge, but `git pull --rebase [tmux-python's remote] master`

Example of what I mean:

`git checkout issue-139`
`git remote add tmux-python`
`git fetch tmux-python` 
`git pull --rebase tmux-python master`
_Resolve any conflicts that arise_
`git push --force-with-lease`

(Assume the branch `issue-139` is already set to target your repo)

**Comment by jerri** [comment]
I see. I will prepare something fresh then.

**Comment by tony** [comment]
@jerri Sorry about that. Thank you! Take your time.

**Comment by tony** [comment]
@jerri Also, have you figured out how to write the tests? e.g. [`test_workspacebuilder.py`](https://github.com/tmux-python/tmuxp/blob/master/tests/test_workspacebuilder.py)

That would definitely make the PR easier to merge

**Comment by jerri** [comment]
Sorry, no testing as of now. First I didn't find a fix for the problem that occured at the time. Second I really didn't find time to do any additional implementation.

**Comment by jerri** [comment]
OK. Rebase has happened.

**Comment by tony** [comment]
@jerri Thank you!

**Comment by tony** [comment]
@jerri I will take a look at if I can add the tests, too

**Comment by tony** [comment]
@jerri If you give me access to your tmuxp repository, I can also rebase it for you and force push it

File: examples/pane-shell.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    suppress_history: false
+    options:
+      remain-on-exit: true
+    panes:
+      - shell: /usr/bin/python3
+        shell_command:
+          - print('This is python 3')
+      - shell: /usr/bin/vim -u none
+        shell_command:
+          - iAll panes have the `remain-on-exit` setting on.
```
Comment (tony) [inline examples/pane-shell.yaml:15]:
```suggestion
          - All panes have the `remain-on-exit` setting on.
```

Comment (tony) [reply examples/pane-shell.yaml:15]: Wait. I'm wrong

Comment (tony) [reply examples/pane-shell.yaml:15]: 😆

Comment (tony) [reply examples/pane-shell.yaml:15]: `i`... _Insert_

I've been in vim every day for 15+ years and this got passed me

**Comment by tony** [comment]
@jerri It's in!

**Comment by tony** [comment]
I will add this into next release, which will hopefully be soon!

**Comment by jerri** [comment]
Awesome. Thanks for the info, and hope it helps. :-D

**Comment by tony** [comment]
@jerri This is live in tmuxp [1.10](https://pypi.org/project/tmuxp/1.10.0/)
